### PR TITLE
pyprland: init at 1.4.1

### PR DESCRIPTION
**Pull request by iliayar**
## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -19850,6 +19850,8 @@ with pkgs;
 
   pqrs = callPackage ../development/tools/pqrs { };
 
+  pyprland = callPackage ../applications/window-managers/hyprwm/pyprland { };
```
Comment (eclairevoyant) [inline pkgs/top-level/all-packages.nix:19853]: please use `pkgs/by-name` hierarchy and remove this

File: pkgs/applications/window-managers/hyprwm/pyprland/default.nix [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{ lib
+, poetry2nix
+, fetchFromGitHub
+}:
+
+
+poetry2nix.mkPoetryApplication rec {
+  projectDir = fetchFromGitHub (lib.importJSON ./src.json);
```
Comment (eclairevoyant) [inline pkgs/applications/window-managers/hyprwm/pyprland/default.nix:8]: why `src.json`? should be inlined

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
Thanks for the contribution! Couple notes, and also please make the adding-to-maintainers-list a separate commit (that comes first). Message should be `maintainers: add iliayar`.

**Comment by MikaelFangel** [comment]
Ref to add maintainer PR: https://github.com/NixOS/nixpkgs/pull/263022

**Comment by eclairevoyant** [comment]
there should not be a separate PR to add maintainer, please combine with this one.

**Comment by MikaelFangel** [comment]
Have closed the add maintainer pr (:

**Comment by eclairevoyant** [comment]
Okay, with all of the fluff out of the way, this way of packaging won't work as there's IFD, which generally isn't allowed in nixpkgs. Python packages in nixpkgs should be written with `buildPythonApplication`/`buildPythonPackage`

**Comment by iliayar** [comment]
Ok, messed up a bit) now should be good

File: pkgs/by-name/py/pyprland/package.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  pname = "pyprland";
+  version = "1.4.1";
+  pyproject = true;
+
+  src = fetchFromGitHub ({
+    owner = "hyprland-community";
+    repo = "pyprland";
+    rev = "02de5fbc7643b52369775060e697c75730448dd7";
+    sha256 = "14gxwxsj261gfyqs938x2gz3bsswrh1j2gchfsb6bdw3cx7ixz0r";
+  });
```
Comment (MikaelFangel) [inline pkgs/by-name/py/pyprland/package.nix:17]:
```suggestion
  src = fetchFromGitHub {
    owner = "hyprland-community";
    repo = "pyprland";
    rev = version;
    hash = "sha256-x/Rar80jwBX64pW+uv0edhlC44OP1b1e2vnJLFGlIms=";
  };
```

File: pkgs/by-name/py/pyprland/package.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{ lib
+, fetchFromGitHub
+, python311
+}:
```
Comment (MikaelFangel) [inline pkgs/by-name/py/pyprland/package.nix:4]: When the list is so short i think it better to write in one line.
```suggestion
{ lib, fetchFromGitHub, python311 }:
```

File: pkgs/by-name/py/pyprland/package.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchFromGitHub ({
+    owner = "hyprland-community";
+    repo = "pyprland";
+    rev = "02de5fbc7643b52369775060e697c75730448dd7";
+    sha256 = "14gxwxsj261gfyqs938x2gz3bsswrh1j2gchfsb6bdw3cx7ixz0r";
+  });
+
+  nativeBuildInputs = with python311.pkgs; [
+    poetry-core
+  ];
```
Comment (MikaelFangel) [inline pkgs/by-name/py/pyprland/package.nix:21]:
```suggestion
  nativeBuildInputs = with python311.pkgs; [ poetry-core ];
```

**Review by MikaelFangel — CHANGES_REQUESTED** [review]

**Review by MikaelFangel — APPROVED** [review]

File: pkgs/by-name/py/pyprland/package.nix [hunk 6]
```diff
@@ -0,0 +1,24 @@
+{ lib, fetchFromGitHub, python311 }:
```
Comment (eclairevoyant) [inline pkgs/by-name/py/pyprland/package.nix:1]:
```suggestion
{ lib, fetchFromGitHub, python3Packages }:
```

File: pkgs/by-name/py/pyprland/package.nix [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+python311.pkgs.buildPythonApplication rec {
```
Comment (eclairevoyant) [inline pkgs/by-name/py/pyprland/package.nix:3]:
```suggestion
python3Packages.buildPythonApplication rec {
```

File: pkgs/by-name/py/pyprland/package.nix [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "hyprland-community";
+    repo = "pyprland";
+    rev = version;
+    hash = "sha256-x/Rar80jwBX64pW+uv0edhlC44OP1b1e2vnJLFGlIms=";
+  };
+
+  nativeBuildInputs = with python311.pkgs; [ poetry-core ];
```
Comment (eclairevoyant) [inline pkgs/by-name/py/pyprland/package.nix:15]:
```suggestion
  nativeBuildInputs = with python3Packages; [ poetry-core ];
```

File: pkgs/by-name/py/pyprland/package.nix [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+  meta = with lib; {
+    description = "An hyperland plugin system";
+    homepage = "https://github.com/hyprland-community/pyprland";
+    license = licenses.mit;
+    maintainers = with maintainers; [ iliayar ];
```
Comment (eclairevoyant) [inline pkgs/by-name/py/pyprland/package.nix:21]:
```suggestion
    maintainers = with maintainers; [ iliayar ];
    mainProgram = "pypr";
```

`meta.mainProgram` should be set when there is a main binary

File: pkgs/by-name/py/pyprland/package.nix [hunk 10]
```diff
@@ -0,0 +1,24 @@
+{ lib, fetchFromGitHub, python311 }:
+
+python311.pkgs.buildPythonApplication rec {
+  pname = "pyprland";
+  version = "1.4.1";
+  pyproject = true;
```
Comment (eclairevoyant) [inline pkgs/by-name/py/pyprland/package.nix:6]:
```suggestion
  format = "pyproject";
```

while your original version works, I would prefer the version that's more common/consistent in nixpkgs.

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Review by eclairevoyant — APPROVED** [review]

**Comment by iliayar** [comment]
Result of `nixpkgs-review pr 263010` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pyprland</li>
    <li>pyprland.dist</li>
  </ul>
</details>

**Comment by pbsds** [comment]
`lib/python3.11/site-packages/pyprland/command.py` is marked executable and has a shebang, but yields the following error when run:

```
Traceback (most recent call last):
  File "/dev/shm/nixpkgs-review/pr-263010/results/pyprland/lib/python3.11/site-packages/pyprland/command.py", line 11, in <module>
    from .common import PyprError, get_logger, init_logger
ImportError: attempted relative import with no known parent package
```

Please `chmod -x` it

**Comment by iliayar** [comment]
I'm no sure where to put it. changing permissions in `postInstall`. The file is not executable now: 
```shell
nixpkgs master* λ ls -la result/lib/python3.11/site-packages/pyprland/command.py
.r--r--r-- 8.1k root  1 Jan  1970 result/lib/python3.11/site-packages/pyprland/command.py
```

File: pkgs/by-name/py/pyprland/package.nix [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    hash = "sha256-x/Rar80jwBX64pW+uv0edhlC44OP1b1e2vnJLFGlIms=";
+  };
+
+  nativeBuildInputs = with python3Packages; [ poetry-core ];
+
+  postInstall = ''
+    # redunant execute permissions on file, that fails when run
+    chmod -x $out/${python3Packages.python.sitePackages}/pyprland/command.py
+  '';
+
```
[the remark this answers is not in the document]
Comment (pbsds) [reply pkgs/by-name/py/pyprland/package.nix:21]: The other ones require `HYPRLAND_INSTANCE_SIGNATURE` to be set

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 263010` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pyprland</li>
    <li>pyprland.dist</li>
  </ul>
</details>


- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on x86_64-linux
- [x] executables tested on x86_64-linux
  - not with hyperland though
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

**Comment by pbsds** [comment]
Welcome to nixpkgs!

**Pull request merged**